### PR TITLE
New stream capture class + startup script to restream twitch streams

### PR DESCRIPTION
**Pull request by timotheeg**
## TLDR
Sort of half baked feature to allow easy restreaming of twitch streams that follow Stencil.

Half baked but working OK!

## Longer story
Should allow a restreamer to host a competition with OCR and a re-renderer, catering for 2 players for now.

PR content:
* New capture class `stream`
* Script `scripts/restream.py`
* NESTrisOCR template file for stencil-based stream

Local requirement as a restreamer: `streamlink`, `ffmpeg`, `vlc`

Script is invoked as:

```bash
python3 -m scripts.restream PLAYER_1_TWITCH_USER_NAME 1

# and for player 2
python3 -m scripts.restream PLAYER_1_TWITCH_USER_NAME 2
```

Script does:
* Locate suitable stream quality (720p30 preferred)
* Create local VLC server from it
* Extract information such as FPS and resolution
* Create player config file based of a template
* invoke NESTrisOCR's calibrator for manual fine tuning and verification (Should be very close, thanks to using the Stencil)
* starts NESTrisOCR with that config

The VLC servers will be accessible from the following 2 URLs (for player 1 and 2 respectively)
* http://localhost:8081
* http://localhost:8082

NESTrisOCR will push its game data to port 4001 and 4002 respectively, where a suitable renderer may be listening (like NESTrisStatsUI 😅 .

This allow for a pretty quick restream setup with sample results as folllow:

image

### CODE REVIEW

File: scripts/restream.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+CAP_RATIOS = [1233 / 1920, 1]  # based on the stencil dimensions for 1080p
+
+PLAYER_SETTINGS = [
+    [8081, 4001],
+    [8082, 4002],
+]
+
+
+def setupPlayer(twitch_name, player_num):
```
Comment (blakegong) [inline scripts/restream.py]: /nit snake_case 😄

File: scripts/restream.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+    m = size_re.search(str(stderr))
+
+    width, height = RESOLUTIONS[m.group("size_key")]
+
+    print("Found size", width, height)
+
+    # Update player config file
+    player_config = (
+        config_base_file.replace("USER_NAME", twitch_name)
```
Comment (blakegong) [inline scripts/restream.py]: This hack is both awesome and eye sore inducing at the same time 😛 JSON with variable is used extensively in Kubernetes, as a "language" called Jsonnet. A Python sample can be found at: https://jsonnet.org/ref/bindings.html#python_api

If you are okay with it, I can issue a subsequent PR later to port your JSON file to `.jsonnet` and update the chunk of code here 😄

**Review by blakegong — APPROVED** [review]
👏

File: scripts/restream.py (hunk 2 above)
Comment (alex-ong) [reply scripts/restream.py]: agree that string-replace is hackerman. Why not just init player_config, then do
```
config["user.user_name"] = twitch_name
config["user.twitch_name"] = twitch_name
config["perf.scan_rate"] = fps
```
etc?

Comment (alex-ong) [reply scripts/restream.py]: Also i've added a commandline hook `--defaultconfig` so you can pass in `config.stream.template.json` into it. 
That might be helpful (or might not be :( )

I could see that chucking `--defaultconfig config.stream.template.json` and `--config config.competition.p{}.json` together, or adding them as args to Config's constructor, and then modifying then calling Config.SaveAs()

**Comment by alex-ong** [comment]
It looks good but the json `stringreplace` is pretty hackerman. Don't worry its a future problem 👍 

Could you 
1) put restreamer stuff inside scripts/restreamer since it might grow
2) add a small readme.md inside that folder?

File: scripts/restream.py (hunk 2 above)
Comment (timotheeg) [reply scripts/restream.py]: Haha, I did consider doing proper config edit (json deserialize, edit local object, json reserialize), but I was looking at the smallest fastest way to get to where I wanted 😅 

Zero objection to using jsonnet instead! :)

**Comment by timotheeg** [comment]
> Could you
> 
> 1. put restreamer stuff inside scripts/restreamer since it might grow
> 2. add a small readme.md inside that folder?

Sure thing, I reckon only the template config file needs to move to `scripts`, and I'll add a readme too

**Comment by alex-ong** [comment]
It's amazing how quickly real programmers work :jealousface:

**Comment by timotheeg** [comment]
I've added a few commits so far, but I'll finish up setting up the config tonight

**Comment by timotheeg** [comment]
PR ready. Some highlights:
* split config into 2 fies: config_class.py an config.py, so that using the config class doens't require adding CLI args
* Bunch of fixes in the restream file. 

Tested!

File: nestris_ocr/config.py [hunk 3]
```diff
@@ -1,155 +1,4 @@
-from cached_property import threaded_cached_property
-from collections import OrderedDict
-import json
-
-from nestris_ocr.utils.sub_image import spawn_subimage
 from nestris_ocr.utils.program_args import args
+from nestris_ocr.config_class import Config
```
Comment (blakegong) [inline nestris_ocr/config.py:2]: `config_class` is a little too un-Python to me 😅 

How about turning `config.py` into a module instead? So:
```
nestris_ocr
  config.py          -> config/__init__.py
  config_class.py    -> config/config.py
```

Edit: Ah nvm, `config.py` would also clash with the `config` variable defined in this file. Let's refactor it later.

File: scripts/restream.py [hunk 4]
```diff
@@ -1,33 +1,28 @@
...
+# entries as [VLC_BRODCAST_PORT, OCR_SEND_PORT]
 PLAYER_SETTINGS = [
     [8081, 4001],
     [8082, 4002],
 ]
 
 
-def setupPlayer(twitch_name, player_num):
-    local_stream_port, ocr_dest_port = PLAYER_SETTINGS[player_num - 1]
+def setupPlayer(twitchName, playerNum):
```
Comment (blakegong) [inline scripts/restream.py]: Why are they camelCase now 😂 My complaint was that things should be in camel_case, so `def setup_player`. In modern Python, we should have either PascalCase (for class names etc.), ALL_CAPS (constants), or camel_case (variable names, function names). 

I should have been more specific to prevent this from happening 😭

File: scripts/restream.py [hunk 5]
```diff
@@ -71,37 +70,46 @@ def setupPlayer(twitch_name, player_num):
...
+    # ==================================
+    # 3. write player config file from template
+
+    playerConfigFile = "config.competition.p{}.json".format(playerNum)
+
+    os.remove(playerConfigFile)
+
+    config = Config(playerConfigFile, auto_save=False, default_config=TEMPLATE_FILE)
+
+    config["player.name"] = twitchName
```
Comment (blakegong) [inline scripts/restream.py]: So clever 👏 The override is a little hidden, but this is such a low cost tweak to the previous hacky version, so very nice!

File: scripts/restream.py (hunk 4 above)
Comment (timotheeg) [reply scripts/restream.py]: LOL! 😅 Fine I'll update all of the names again tomorrow

File: nestris_ocr/config.py (hunk 3 above)
Comment (timotheeg) [reply nestris_ocr/config.py:2]: Yeah naming was a pita 😔. Better suggestions welcome.

File: scripts/restream.py (hunk 4 above)
Comment (alex-ong) [reply scripts/restream.py]: Yeah its 99% my fault for having camelCase in the original code base. Blake has graciously helped rename a lot of stuff, but we want to move towards what he said. The code is a mish-mash atm and we want new code to comply.

File: scripts/restream.py (hunk 5 above)
Comment (alex-ong) [reply scripts/restream.py]: :clap::clap::clap:

File: nestris_ocr/capturing/stream.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+    def get_image(self, rgb: bool = False) -> Tuple[float, Image.Image]:
+        if not self.cv2_retval:
+            raise Exception("Faulty capturing device")
+
+        with self.read_lock:
+            cv2_image = self.frame_buffer.popleft()
+
+        if rgb:
+            # can we do this in the thread pool?
```
Comment (alex-ong) [inline nestris_ocr/capturing/stream.py]: Answered in discord; but since we are using thread (not process) it doesnt matter too much where we RGB it. Also if the stream is h264, it might *already* be rgb output (though its funneling through opencv, so its probably BGR). You could try displaying the image with rgb=false, and if its rgb you can remove these lines, if its bgr keep them. Not a big deal for this PR at all, i wouldn't bother testing it till later.

File: main.py [hunk 7]
```diff
@@ -34,14 +35,14 @@ def main(on_cap, check_network_close):
             time.sleep(RATE)
             continue
 
-        frame_end_ts = ts + RATE
+        frame_end_ts = (ts or read_ts) + RATE
```
Comment (alex-ong) [inline main.py:38]: > (ts or read_ts)

This is fine for now but i'd rather that we pass in speed into stream.py instead. In the future i would like to have source_fps and ocr_fps separated; where both can have automatic options:

* Capture cards present FPS, so with AUTO we can set it to capture_card.fps * [2 if (interlace != discard)]
* OCR_FPS should be 30 or 60 by default, with using AUTO to first check what the capture method's FPS is, followed by dynamically scaling fps from lagtime (e.g. peasant computer can OCR at 10fps to get a smoother output)

So for after this PR, either:
* add `capture_fps` into config, rename `performance.scan_rate` to `ocr_fps` into config, then make stream.py respect it (older classes can ignore it, that's a future problem) and main.py use ocr_fps
* more hacky: just make stream.py read the config and use `performance.scan_rate`


But for now lets just leave it unless you love doing more work :D

File: scripts/restream.py (hunk 4 above)
Comment (timotheeg) [reply scripts/restream.py]: All names following in script following correct convention now 👍 🎉

File: scripts/restream.py (hunk 5 above)
Comment (timotheeg) [reply scripts/restream.py]: That very recent addition of `default_config` was great to use for this case! :D

File: nestris_ocr/config.py (hunk 3 above)
Comment (alex-ong) [reply nestris_ocr/config.py:2]: I dont like the new names but i cant think of a better solution 👍

File: main.py (hunk 7 above)
Comment (timotheeg) [reply main.py:38]: Noted! Let's merge this for now since it works, and we can improve in future PR. 

Oh btw, using different `source_fps` and `ocr_fps` might require extra changes to stream.py 🤔 . Otherwise, at the beginning of the game, until the frame buffer fills up, ocr would read continuous frames but at a "slow motion rate". It should stabilize once the frame buffer is full (dequeue will drop the frames), but the beginning of capture might look weird.

Comment (alex-ong) [reply main.py:38]: Yep lets not worry about it for now since its *not a trivial 5 line change*.

I would assume stream.py would simply return blank or the first frame for the first [buffer_size] frames, and not remove frames from the buffer until it is say 10~ long. After a comfortable buffer has been created, it would behave normally. This does mean a 10/30 == 1/3rd second delay but that's pretty good.

**Comment by timotheeg** [comment]
K! I think it's ready :)

I can literally be up and running streaming a match in 1 minute now 💪 

You can test it too by following these steps:

pre-req:
clone [NESTrisStatsUI](https://github.com/timotheeg/NESTrisStatsUI) + npm install in it

1. open 3 terminals windows
2. In first window, go to NEStrisStatsUI base folder and run `node app_competition.js`
3. In a browser tab, open local html file `NESTrisStatsUI/public/competition_layout1_ctjc.html`
4. visit https://www.twitch.tv/directory/game/Tetris, and locate 2 twitch players streaming NTSC tetris with stencil with reasonably good stream quality. Note their twitch names down, assumed here as "twitch1" and "twitch2"
5. In second terminal window, run `python3 -m scripts.restream twitch1 1`, fine tune when calibrator shows (if needed)
6. In third terminal window, run `python3 -m scripts.restream twitch2 2`, fine tune when calibrator shows (if needed) (you can run this in parallel to step 5)
7. check the browser tab. Booya! competition is ongoing with score differential, simple stats, and crisp pixel-perfect rendering 💪 

Now the sad part, a bunch of processes (VLC local servers) run in the background and there's no admin UI to kill them. So when I want to stop the processes, I run this on mac:
```
killall python python3 python3.8 vlc Python node
```

I don't know what the equivalent would be in windows 😅 

Last few Notes: 
1. to control the competition interface (player names, number of hearts to Victory), in a separate browser tab, open `NESTrisStatsUI/public/competition_admin.html` (If you restart the node server, you need to refresh this page, so the websocket connects to the server again)
2. To stream, in OBS/StreamLabs, set up to stream at 720p, create a new Scene, add a Browser source, make it cover the entire scene at 1280x720, make it load `NESTrisStatsUI/public/competition_layout1_ctjc.html` (just like the admin UI, if you restart the server, the scene must be refreshed so it connects to the correct node server.

**Comment by alex-ong** [comment]
> Now the sad part, a bunch of processes (VLC local servers) run in the background and there's no admin UI to kill them. So when I want to stop the processes, I run this on mac:
> 
> ```
> killall python python3 python3.8 vlc Python node
> ```
You could use `processes.append(StartProcess(args))` as you already have and then have a simple tkinter button that says kill all, which would call `process.abort()`. Needs minor re-architecting, and wouldnt kill `node` but should kill vlc/python assuming you started it from here

**Comment by timotheeg** [comment]
Nice suggestion! I'll open another PR when I have time to do this. Would indeed be a lot nicer if anyone actually uses this for restreaming :)

File: scripts/restream.py (hunk 2 above)
Comment (timotheeg) [reply scripts/restream.py]: Cleaned up with using the `Config` class 👍

File: main.py (hunk 7 above)
Comment (timotheeg) [reply main.py:38]: Hmm, If we wait for buffer to be filled, then the delay should be related to the buffer size, not the ratio of read and scan rates 🤔  Delay of stream would be be (buffer_size / source_fps). With the current (completely arbitrary)  buffer size of 150, a 30fps stream would be delayed by 5s before we start ocr/rendering.

**Comment by alex-ong** [comment]
> Nice suggestion! I'll open another PR when I have time to do this. Would indeed be a lot nicer if anyone actually uses this for restreaming :)

I think sharky is interested, and i think he's technical enough to command line a few things.

**Comment by timotheeg** [comment]
Yeah, I had pinged him to ask if he waned to give a try. Now that this is merged, I'll help him set up. Exciting times! :D

File: main.py (hunk 7 above)
Comment (alex-ong) [reply main.py:38]: > ` If we wait for buffer to be filled`

The `fill before allowing frame ejection` could be say 20-30 and the full buffer could be 150. then if the buffer gets completely empty we reset it to not remove items from buffer till it hits 20-30 again. This is similar to how streaming video in the first place works:
1) pause until buffer is at least 1s (return buffer[0], don't pop())
2) (loop) receive from internet + play at realtime rate  (buffer[0].pop(), return popped)
3) if buffer hits 0, go to step 1

This assumes an unbounded buffer. If you have correct timestamps and we implement ocr_rate correctly (its not currently correct), an ocr_rate of 10 and stream_rate of 30 would mean ocr just skips 3 frames every scan and realtime still stays "correct". It would 100% wreak havoc on any server side stuff that doesnt want skipped frames though, but if the ocr isn't fast enough because peasant computer, real-time is probably more important than having a growing stream lag.

It's getting more and more complicated, but then the server would have to support determining if the computer is "dropping frames" or not and disabling effects like field-based tetris detection, since that doesnt work if you're dropping frames like crazy.
